text/0000-delete-prefix.md

```diff
@@ -0,0 +1,135 @@
+used. This solution could work but will slow down executing of every regular 
+transactions while the clean transaction is rare.
+
+No change is recommended to the existing methods, so the implementation of this
+RFC should not cause any breakage for pre-existing transaction processors.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Add new API to context: deletePrefix that delete a subtree of addresses 
+starting from a given address prefix. 
+This API should be similar to regular delete address API but can get as input
+prefix of address and not just full 35 bytes address.
+Calling this API with full address should behave exactly like calling existing
+delete API.
+
+In the suggested implementation there are two proto changes:
+
+Addition to state_context.proto:
+	
+	// A request from the handler/tp to delete state prefix entries at an 
+	// collection of addresses prefix subtree
+	message TpStateDeletePrefixRequest {
+		string context_id = 1;
+		repeated string addresses_prefix = 2;
+	}
+
+	// A response form the contextmanager/validator with the addresses that 
+	// were deleted
+	message TpStateDeleteResponse {
+		enum Status {
+			STATUS_UNSET = 0;
+			OK = 1;
+			AUTHORIZATION_ERROR = 2;
+		}
+
+		repeated string addresses = 1;
+		Status status = 2;
+	}
+
+Addition to validator.proto:
+
+	// State delete prefix request from the transaction processor to the 
+	// validator/context_manager
+	TP_STATE_DELETE_PREFIX_REQUEST = 17;
+	// State delete prefix response from the validator/context_manager to the
+	// transaction processor
+	TP_STATE_DELETE_PREFIX_RESPONSE = 18;
+	
+	
+# Drawbacks
+[drawbacks]: #drawbacks
+
+1. This will need to be implemented in every sawtooth SDK.
+2. Once implemented it will be difficult to deprecate because of our 
+   commitment to backward compatibility.
+3. This requires transaction processor developers to understand the risk of
+   calling delete by prefix and accidently deleting unintended addresses.
+   Delete by prefix should be used only when developer is sure prefix is 
+   unique like when having a pre-defined prefix and not in case there could
+   be hash collisions for the prefix.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+An alternative could be in the transaction processor logic, store an address
+that contains list of existing addresses. When need to delete multiple 
+addresses transaction processor will first read the special address, then 
+delete each address in the list.
+This has a performance downside that each time transaction processor needs to 
+write a new address, it will need to modify the special address containing 
+list of all addresses. This address could be huge and it needs to be changed
+for almost every transaction, while the need to delete all the addresses in 
+the list is very rare.
+
+# Prior art
+[prior-art]: #prior-art
+
+Access address by prefix is a key capability of radix tree.
```

Comment:
This is worded badly. In fact, I don't know that I would go so far as to even include prior art, since it is an extension of current Sawtooth - the Merkle-Radix trie - which that functionality made use of prior art.